---
title: "HW4_gloria_chi"
output: html_document
---
## Assignment 4
This is your fourth and final homework due March 12.

You will have to analyze the RNA-seq data presented in: Henn, A. D. et al. High-resolution temporal response patterns to influenza vaccine reveal a distinct human plasma cell gene signature. Scientific Reports 3, 2327 (2013).

1. Get the data from GEO. Please look at the class lecture slides as we've done it already
2. Use voom and limma to find genes that are differentially expressed at each time point compared to baseline (day 0). Use an FDR cutoff of 0.01. Display your results using pheatmap showing the log fold-change of the differentially expressed genes grouped by time point.
3. Perform a GSEA analysis using camera and the MSigDB Reactome pathway gene signatures. Display your results using pheatmap, again group by timepoint. This is similar to what we've done in class.

## Setup `knitr`

Set some global knitr options

```{r, cache=FALSE}
library("knitr")
opts_chunk$set(tidy=FALSE, cache=TRUE, messages=FALSE)
```

##Load libraries

```{r}
library(GEOquery)
library(Biobase)
library(data.table)
library(limma)
library(pheatmap)
library(GSEABase)
```

## Get data

Get GEO dataset

```{r}
gd <- getGEO("GSE45735", destdir = "./")
pd <- pData(gd[[1]])
getGEOSuppFiles("GSE45735", makeDirectory = FALSE, baseDir = "./")

# Note the regular expression to grep file names
files <- list.files(path = "./", pattern = "GSE45735_T.*.gz", 
    full.names = TRUE)

# Read in gzip-compressed, tab-delimited files
file_list <- lapply(files, read.table, sep = "\t", header = TRUE)

# Subset to only those rows where Gene contains only non-space characters
# This addresses problems with T14 file containing 28 invalid rows at end of file
file_list <- lapply(file_list, function(file_list)subset(file_list, grepl('^[^[:space:]]+$', Gene)))

# Remove duplicated rows
file_list_unique <- lapply(file_list, function(x){x<-x[!duplicated(x$Gene),]; 
                                                  x <- x[order(x$Gene),]; 
                                                  rownames(x) <- x$Gene;
                                                  x[,-1]})
# Take the intersection of all genes
gene_list <- Reduce(intersect, lapply(file_list_unique, rownames))
file_list_unique <- lapply(file_list_unique, "[", gene_list,)
matrix <- as.matrix(do.call(cbind, file_list_unique))

# Clean up the pData
pd_small <- pd[!grepl("T13_Day8",pd$title),]
pd_small$Day <- sapply(strsplit(gsub(" \\[PBMC\\]", "", pd_small$title),"_"),"[",2)
pd_small$subject <- sapply(strsplit(gsub(" \\[PBMC\\]", "", pd_small$title),"_"),"[",1)
colnames(matrix) <- rownames(pd_small)
```
Create an eSet we can use:

```{r}
# Note that I add one to the count
new_set <- ExpressionSet(assayData = matrix+1)
pData(new_set) <- pd_small
```
Set-up our design matrix to estimate our weights:

```{r}
design <- model.matrix(~subject+Day, new_set)
new_set_voom <- voom(new_set,design = design)
```

##Differential Expression Analysis with Voom
```{r}
lm <- lmFit(new_set_voom, design)
eb <- eBayes(lm)
# Look at all days compared to Day 0
tt_results <- list()
sig_genes <- list()
for (i in 1:10){
tt <- topTable(eb, coef = paste0("DayDay",i), number = Inf, adjust.method="BH", sort.by="none")
tt_results[[i]] <- tt
sig_genes[[i]] <- rownames(tt[tt$adj.P.Val < 0.01,])  
}
```

##Visualize Data - Heatmap of Differential Gene Expression for Days 1-10

```{r}
#Subset the tt matrices to only significant genes
tt_sub <- list()
tt_sub <- lapply(tt_results, function(x) (x[rownames(x) %in% sig_genes[[1]] | rownames(matrix) %in% sig_genes[[3]],]))

Day1 <- tt_sub[[1]]$logFC
Day2 <- tt_sub[[2]]$logFC
Day3 <- tt_sub[[3]]$logFC
Day4 <- tt_sub[[4]]$logFC
Day5 <- tt_sub[[5]]$logFC
Day6 <- tt_sub[[6]]$logFC
Day7 <- tt_sub[[7]]$logFC
Day8 <- tt_sub[[8]]$logFC
Day9 <- tt_sub[[9]]$logFC
Day10 <- tt_sub[[10]]$logFC

hm_matrix <- cbind(Day1, Day2, Day3, Day4, Day5, Day6, Day7, Day8, Day9, Day10)
row.names(hm_matrix) <- row.names(tt_sub[[1]])

pheatmap(hm_matrix, cluster_cols=FALSE,fontsize_row = 5)
```

##Using CAMERA

Convert the gene sets to gene indices
```{r}
#Only get the reactome pathway indices from MSigDB
#requires registering on their website and downloading appropriate file
c2_set <- getGmt("./c2.cp.reactome.v4.0.symbols.gmt")
gene_ids <- geneIds(c2_set)

# Camera requires gene-indices. 
# Which function to use will depend on which version of limma you have.
#     http://bioconductor.org/packages/release/bioc/news/limma/NEWS
#     "symbols2indices() renamed to ids2indices()."
if (exists("ids2indices")) { 
    sets_indices <- ids2indices(gene_ids, rownames(new_set))
}
if (exists("symbols2indices")) {
    sets_indices <- symbols2indices(gene_ids, rownames(new_set))    
}
```

##Finding enriched gene sets

As with limma, we need to specify the contrast we wish to test at the set level:
```{r}
# Note that camera works on voom objects
cont_matrix <- makeContrasts("DayDay1", levels=design)
res <- camera(new_set_voom, sets_indices, design=design, cont_matrix)
res[1:10, ]
```
##Finding enriched gene sets over time
```{r}
res <- vector("list",length = 10)
for(i in 1:10)
  {
  contrast <- paste0("DayDay",i)
  cont_matrix <- makeContrasts(contrast, levels=design)
  res[[i]] <- camera(new_set_voom, sets_indices, design=design, contrast=cont_matrix, sort=FALSE)
}
```

##Visualizing the results - Heatmap of Enriched Reactome Pathways
```{r}
PValue <- sapply(res, function(x){ifelse(x$Direction=="Up", -10*log10(x$PValue), 10*log10(x$PValue))})
rownames(PValue) <- rownames(res[[1]])
PValue_max <- rowMax(abs(PValue))
PValue_small <- PValue[PValue_max>30, ]
anno <- data.frame(Time=paste0("Day",1:10))
rownames(anno) <- colnames(PValue_small)  <- paste0("Day",1:10)

pheatmap(PValue_small, cluster_cols=FALSE,fontsize_row = 5)
```

